File: Java-Core/src/main/java/presto/statement/CollectionSwitchCase.java
package presto.statement;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.type.CollectionType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.ListValue;


public class CollectionSwitchCase extends SwitchCase {

	public CollectionSwitchCase(IExpression expression, StatementList list) {
		super(expression,list);
	}

	@Override
	public void checkSwitchType(Context context, IType type) throws SyntaxError {
		IType thisType = expression.check(context);
		if(thisType instanceof CollectionType)
			thisType = ((CollectionType)thisType).getItemType();
		if(!thisType.isAssignableTo(context, type))
			throw new SyntaxError("Cannot assign:" + thisType.getName() + " to:" + type.getName());
	}
	
	@Override
	public boolean matches(Context context, IValue value) throws PrestoError {
		Object thisValue = expression.interpret(context);
		if(thisValue instanceof ICollection)
			return ((ICollection<?>)thisValue).hasItem(context, value);
		else if(thisValue instanceof ListValue)
			return ((ListValue)thisValue).hasItem(context, value);
		return false;
	}
	
	@Override
	public void caseToPDialect(CodeWriter writer) {
		caseToEDialect(writer);
	}
	
	@Override
	public void caseToODialect(CodeWriter writer) {
		writer.append("case in ");
		expression.toDialect(writer);
		writer.append(":\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
	}
	
	@Override
	public void caseToEDialect(CodeWriter writer) {
		writer.append("when in ");
		expression.toDialect(writer);
		writer.append(":\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
	}
	
	@Override
	public void catchToODialect(CodeWriter writer) {
		writer.append("catch (");
		expression.toDialect(writer);
		writer.append(") {\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
		writer.append("} ");
	}
	
	@Override
	public void catchToPDialect(CodeWriter writer) {
		writer.append("except in ");
		expression.toDialect(writer);
		writer.append(":\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
	}
	
	@Override
	public void catchToEDialect(CodeWriter writer) {
		caseToEDialect(writer); // no difference
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/statement/CollectionSwitchCase.js
var CollectionType = require("../type/CollectionType").CollectionType;
var SwitchCase = require("./SwitchCase").SwitchCase;

function CollectionSwitchCase(expression, instructions) {
	SwitchCase.call(this, expression,instructions);
	return this;
};

CollectionSwitchCase.prototype = Object.create(SwitchCase.prototype);
CollectionSwitchCase.prototype.constructor = CollectionSwitchCase;

CollectionSwitchCase.prototype.checkSwitchType = function(context, type) {
	var thisType = this.expression.check(context);
	if(thisType instanceof CollectionType) {
		thisType = thisType.itemType;
	}
	if(!thisType.isAssignableTo(context, type)) {
		throw new SyntaxError("Cannot assign:" + thisType.name + " to:" + type.name);
	}
};

CollectionSwitchCase.prototype.matches = function(context, value) {
	var thisValue = this.expression.interpret(context);
	if(thisValue.hasItem) {
		return thisValue.hasItem(context, value);
	} else {
		return false;
	}
};


CollectionSwitchCase.prototype.caseToPDialect = function(writer) {
    this.caseToEDialect(writer);
}


CollectionSwitchCase.prototype.caseToODialect = function(writer) {
    writer.append("case in ");
    this.expression.toDialect(writer);
    writer.append(":\n");
    writer.indent();
    this.statements.toDialect(writer);
    writer.dedent();
}


CollectionSwitchCase.prototype.caseToEDialect = function(writer) {
    writer.append("when in ");
    this.expression.toDialect(writer);
    writer.append(":\n");
    writer.indent();
    this.statements.toDialect(writer);
    writer.dedent();
}


CollectionSwitchCase.prototype.catchToODialect = function(writer) {
    writer.append("catch (");
    this.expression.toDialect(writer);
    writer.append(") {\n");
    writer.indent();
    this.statements.toDialect(writer);
    writer.dedent();
    writer.append("} ");
}


CollectionSwitchCase.prototype.catchToPDialect = function(writer) {
    writer.append("except in ");
    expression.toDialect(writer);
    writer.append(":\n");
    writer.indent();
    statements.toDialect(writer);
    writer.dedent();
}


CollectionSwitchCase.prototype.catchToEDialect = function(writer) {
    caseToEDialect(writer); // no difference
}

exports.CollectionSwitchCase = CollectionSwitchCase;
